okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java

```diff
@@ -96,6 +96,10 @@ public void transportHeadersReceived(List<Header> headers, boolean endOfStream)
     synchronized (lock) {
       if (endOfStream) {
         transportTrailersReceived(Utils.convertTrailers(headers));
+        if (getTransportError() != null) {
```

Comment:
I'm not comfortable with this. You are requiring the two pieces of code stay in sync, and it isn't obvious that one depends on the other's behavior.

Instead, can we just have transportHeadersReceieved always send cancel if there was an error? Simply removing the endOfStream condition seems the simplest and most maintainable solution.